File: Satellite_Architecture/Satellite_Control_Simulation/cpp/src/ode4m.cpp
// ode4m.cpp
#include "ode4m.h"
#include "eulerseqns2.h" // provides extern std::array<double,3> L and function eulerseqns2(...)
#include "iss_orbit.h"   // issOrbit
#include "igrf.h"        // igrf
#include "ned2eci.h"     // ned2eci

#include <cmath>
#include <stdexcept>
#include <array>
#include <vector>
#include <algorithm>

enum { NX = 7 }; // state dimension: [w1,w2,w3,q1,q2,q3,q4]

// Global history vectors for plotting/analysis:
std::vector<std::array<double,3>> LHIST;  // control torque history (body‐frame)
std::vector<std::array<double,3>> MHIST;  // dipole moment history (body‐frame)

std::vector<std::array<double,7>> ode4m(
    const std::function<std::array<double,7>(double, const std::array<double,7>&)> &odefun,
    const std::vector<double> &tspan,
    const std::array<double,7> &y0,
    int cmode
) {
    size_t N = tspan.size();
    if (N < 2) {
        throw std::invalid_argument("tspan must have at least two entries");
    }
    // Step sizes h[i] = tspan[i] - tspan[i-1]
    std::vector<double> h(N);
    for (size_t i = 1; i < N; ++i) {
        h[i] = tspan[i] - tspan[i-1];
        if (h[i] <= 0) {
            throw std::invalid_argument("tspan entries must be strictly increasing");
        }
    }
    // Pre-allocate solution array: Y[i] is state at tspan[i]
    std::vector<std::array<double,7>> Y(N);
    Y[0] = y0;

    // Initialize history arrays
    LHIST.assign(N, {0.0,0.0,0.0});
    MHIST.assign(N, {0.0,0.0,0.0});

    // Pre-simulate Earth's magnetic field at 1 Hz for 16001 seconds
    const size_t M = 16001;
    std::vector<double> Bx_i(M), By_i(M), Bz_i(M);
    for (size_t step = 0; step < M; ++step) {
        double dt = static_cast<double>(step);
        double lat, lon, alt;
        issOrbit(dt, lat, lon, alt); // lat,lon in rad, alt in km
        // IGRF wants lat/lon in degrees, altitude in km
        double Bxn, Byn, Bzn;
        igrf(dt,                   // time (we could pass real epoch here if needed)
             lat * 180.0/M_PI,     // geodetic latitude [deg]
             lon * 180.0/M_PI,     // geodetic longitude [deg]
             alt,                  // altitude [km]
             true,                 // geodetic flag
             Bxn, Byn, Bzn);       // output in nT, in NED
        std::array<double,3> ned = { Bxn, Byn, Bzn };
        auto eci = ned2eci(ned,
                          lat * 180.0/M_PI,
                          lon * 180.0/M_PI,
                          dt);
        Bx_i[step] = eci[0];
        By_i[step] = eci[1];
        Bz_i[step] = eci[2];
    }

    // Spacecraft inertia properties (same rectangular prism)
    double mass = 10.0; // kg
    double height = 0.34, width = 0.20, depth = 0.10; // m
    double Jx = mass/12.0 * (width*width + depth*depth);
    double Jy = mass/12.0 * (height*height + depth*depth);
    double Jz = mass/12.0 * (height*height + width*width);
    double d1 = 1.0 / Jx;
    double d2 = 1.0 / Jy;
    double d3 = 1.0 / Jz;

    // Control parameters
    const double moment = 0.2;     // A·m² (max dipole)
    const double sample_rate = 1.0; // sec

    // Variables for B-dot or ω×B logic
    double tlast = 0.0;
    double Bxlast = 0.0, Bylast = 0.0, Bzlast = 0.0;
    double mxlast = 0.0, mylast = 0.0, mzlast = 0.0;

    // Temporary arrays for RK4 and rotation
    std::array<double,7> F1, F2, F3, F4;
    std::array<double,7> yi, ytemp;

    for (size_t i = 1; i < N; ++i) {
        double ti = tspan[i-1];
        double hi = h[i];
        yi = Y[i-1];

        // 1) Extract angular rates & quaternion from yi
        double w1 = yi[0], w2 = yi[1], w3 = yi[2];
        double q1 = yi[3], q2 = yi[4], q3 = yi[5], q4 = yi[6];

        // Build body←ECI DCM from quaternion:
        double C11 = 1 - 2*(q2*q2 + q3*q3);
        double C12 = 2*(q1*q2 + q3*q4);
        double C13 = 2*(q3*q1 - q2*q4);
        double C21 = 2*(q2*q1 - q3*q4);
        double C22 = 1 - 2*(q3*q3 + q1*q1);
        double C23 = 2*(q2*q3 + q1*q4);
        double C31 = 2*(q3*q1 + q2*q4);
        double C32 = 2*(q2*q3 - q1*q4);
        double C33 = 1 - 2*(q1*q1 + q2*q2);

        // 2) Interpolate Earth‐field at time ti (linear interp if ti not integer)
        double idx_f = std::floor(ti);
        size_t i0 = (idx_f < 0) ? 0 : static_cast<size_t>(idx_f);
        if (i0 >= M-1) i0 = M-1;
        size_t i1 = std::min(i0 + 1, M-1);
        double frac = ti - std::floor(ti);
        double Bxeci = (1 - frac)*Bx_i[i0] + frac*Bx_i[i1];
        double Byeci = (1 - frac)*By_i[i0] + frac*By_i[i1];
        double Bzeci = (1 - frac)*Bz_i[i0] + frac*Bz_i[i1];

        // 3) Rotate ECI‐field into body frame: B_body = DCM_body←ECI * B_ECI
        double Bx_body = C11*Bxeci + C12*Byeci + C13*Bzeci;
        double By_body = C21*Bxeci + C22*Byeci + C23*Bzeci;
        double Bz_body = C31*Bxeci + C32*Byeci + C33*Bzeci;
        // Convert from nT → Tesla
        std::array<double,3> B_body_vec = { Bx_body*1e-9,
                                            By_body*1e-9,
                                            Bz_body*1e-9 };

        // 4) Compute control dipole (A·m²) based on cmode
        std::array<double,3> mvec_body = {0.0, 0.0, 0.0};

        if (cmode == 1) {
            // ---- B-dot (bang-bang) control ----
            if (std::fmod(ti, sample_rate) < 1e-6) {
                if (std::fabs(ti - tlast) > 1e-8) {
                    double Bdotx = (Bx_body - Bxlast) / (ti - tlast);
                    double Bdoty = (By_body - Bylast) / (ti - tlast);
                    double Bdotz = (Bz_body - Bzlast) / (ti - tlast);
                    mvec_body = {
                        -moment * ((Bdotx > 0) - (Bdotx < 0)),
                        -moment * ((Bdoty > 0) - (Bdoty < 0)),
                        -moment * ((Bdotz > 0) - (Bdotz < 0))
                    };
                    Bxlast = Bx_body;
                    Bylast = By_body;
                    Bzlast = Bz_body;
                    tlast = ti;
                } else {
                    // First sample at t=0
                    mvec_body = {0.0, 0.0, 0.0};
                    Bxlast = Bx_body;
                    Bylast = By_body;
                    Bzlast = Bz_body;
                    tlast = ti;
                }
            } else {
                // Hold previous m
                mvec_body = { mxlast, mylast, mzlast };
            }
        }
        else if (cmode == 2) {
            // ---- ω × B “damping” control ----
            if (std::fmod(ti, sample_rate) < 1e-6) {
                if (std::fabs(ti - tlast) > 1e-8) {
                    double Bnorm = std::sqrt(
                        B_body_vec[0]*B_body_vec[0] +
                        B_body_vec[1]*B_body_vec[1] +
                        B_body_vec[2]*B_body_vec[2]
                    );
                    // Normalized field unit vector
                    std::array<double,3> b_unit = {
                        B_body_vec[0]/Bnorm,
                        B_body_vec[1]/Bnorm,
                        B_body_vec[2]/Bnorm
                    };
                    // Cross product w × b
                    std::array<double,3> w_vec = { w1, w2, w3 };
                    std::array<double,3> cross_wb = {
                        w_vec[1]*b_unit[2] - w_vec[2]*b_unit[1],
                        w_vec[2]*b_unit[0] - w_vec[0]*b_unit[2],
                        w_vec[0]*b_unit[1] - w_vec[1]*b_unit[0]
                    };
                    double k = 1.0; // scaling factor
                    mvec_body = {
                        (k / Bnorm) * cross_wb[0],
                        (k / Bnorm) * cross_wb[1],
                        (k / Bnorm) * cross_wb[2]
                    };
                    // Clip each component to ±0.2 A·m²
                    for (int c=0; c<3; ++c) {
                        if (mvec_body[c] > moment)       mvec_body[c] = moment;
                        else if (mvec_body[c] < -moment) mvec_body[c] = -moment;
                    }
                    Bxlast = Bx_body;
                    Bylast = By_body;
                    Bzlast = Bz_body;
                    tlast = ti;
                } else {
                    // First sample at t=0
                    mvec_body = {0.0, 0.0, 0.0};
                    Bxlast = Bx_body;
                    Bylast = By_body;
                    Bzlast = Bz_body;
                    tlast = ti;
                }
            } else {
                // Hold previous m
                mvec_body = { mxlast, mylast, mzlast };
            }
        }

        // Remember this dipole for the next iteration
        mxlast = mvec_body[0];
        mylast = mvec_body[1];
        mzlast = mvec_body[2];

        // 5) Compute control torque in body frame: τ = m × B
        std::array<double,3> torque_body = {
            mvec_body[1]*B_body_vec[2] - mvec_body[2]*B_body_vec[1],
            mvec_body[2]*B_body_vec[0] - mvec_body[0]*B_body_vec[2],
            mvec_body[0]*B_body_vec[1] - mvec_body[1]*B_body_vec[0]
        };
        // Scale by 1/J to feed into Euler’s eqn
        L[0] = torque_body[0] * d1;
        L[1] = torque_body[1] * d2;
        L[2] = torque_body[2] * d3;

        // Store into history arrays
        LHIST[i] = { L[0], L[1], L[2] };
        MHIST[i] = mvec_body;

        // 6) Classical RK4 (call eulerseqns2 via the lambda 'odefun'):
        F1 = odefun(ti, yi);

        for (int j = 0; j < NX; ++j) {
            ytemp[j] = yi[j] + 0.5 * hi * F1[j];
        }
        F2 = odefun(ti + 0.5*hi, ytemp);

        for (int j = 0; j < NX; ++j) {
            ytemp[j] = yi[j] + 0.5 * hi * F2[j];
        }
        F3 = odefun(ti + 0.5*hi, ytemp);

        for (int j = 0; j < NX; ++j) {
            ytemp[j] = yi[j] + hi * F3[j];
        }
        F4 = odefun(tspan[i], ytemp);

        for (int j = 0; j < NX; ++j) {
            Y[i][j] = yi[j]
                    + (hi / 6.0) * (F1[j] + 2.0*F2[j] + 2.0*F3[j] + F4[j]);
        }

        // 7) **Normalize quaternion** so ‖q‖ = 1
        double rq1 = Y[i][3], rq2 = Y[i][4], rq3 = Y[i][5], rq4 = Y[i][6];
        double normq = std::sqrt(rq1*rq1 + rq2*rq2 + rq3*rq3 + rq4*rq4);
        Y[i][3] = rq1 / normq;
        Y[i][4] = rq2 / normq;
        Y[i][5] = rq3 / normq;
        Y[i][6] = rq4 / normq;
    }

    return Y;
}

File: Satellite_LEO_Magnetorquer_Despin_Controller/ode4m.m
function Y = ode4m(odefun,tspan,y0,cmode,varargin)
% ODE4M modified ODE4 fixed step size integrator
% AE 6356 Spacecraft Attitude
%
% Solve differential equations with a non-adaptive method of order 4.
%   Y = ODE4(ODEFUN,TSPAN,Y0) with TSPAN = [T1, T2, T3, ... TN] integrates 
%   the system of differential equations y' = f(t,y) by stepping from T0 to 
%   T1 to TN. Function ODEFUN(T,Y) must return f(t,y) in a column vector.
%   The vector Y0 is the initial conditions at T0. Each row in the solution 
%   array Y corresponds to a time specified in TSPAN.
%
%   Y = ODE4(ODEFUN,TSPAN,Y0,P1,P2...) passes the additional parameters 
%   P1,P2... to the derivative function as ODEFUN(T,Y,P1,P2...). 
%
%   This is a non-adaptive solver. The step sequence is determined by TSPAN
%   but the derivative function ODEFUN is evaluated multiple times per step.
%   The solver implements the classical Runge-Kutta method of order 4.   
%
%   Example 
%         tspan = 0:0.1:20;
%         y = ode4(@vdp1,tspan,[2 0]);  
%         plot(tspan,y(:,1));
%     solves the system y' = vdp1(t,y) with a constant step size of 0.1, 
%     and plots the first component of the solution.   
%
%% set global variables
global L LHIST MHIST % global variables control torque vector and torque history

L=zeros(3,1); % L=[L1;L2;L3] total control torque per axis
%% check inputs
if ~isnumeric(tspan)
  error('TSPAN should be a vector of integration steps.');
end

if ~isnumeric(y0)
  error('Y0 should be a vector of initial conditions.');
end

h = diff(tspan);
if any(sign(h(1))*h <= 0)
  error('Entries of TSPAN are not in order.') 
end  

try
  f0 = feval(odefun,tspan(1),y0,varargin{:});
catch
  msg = ['Unable to evaluate the ODEFUN at t0,y0. ',lasterr];
  error(msg);  
end  

y0 = y0(:);   % Make a column vector.
if ~isequal(size(y0),size(f0))
  error('Inconsistent sizes of Y0 and f(t0,y0).');
end  

%% initialize params
neq = length(y0);
N = length(tspan);
Y = zeros(neq,N);
F = zeros(neq,4);
Y(:,1) = y0;

%% specify control mode
% cmode=2; % 1 = Bdot control, 2 = wxb control

%% initialize global variables
LHIST=zeros(3,N);   % history of control torque vectors
MHIST=zeros(3,N);   % history of magnetic dipole vectors

%% simulate Earth's magnetic field
Bx_ECI = zeros(16001,1);
By_ECI = zeros(16001,1);
Bz_ECI = zeros(16001,1);
B_ECI_mag = zeros(16001,1);
timey_wimey = zeros(16001,1);

stp = 0;
for dt = [0:1:16000]
    stp = stp + 1;
    
    [latitude,longitude,altitude]=issorb(dt); %rad, rad, km
    
    time = 739526;
    [Bx_North, By_East, Bz_Down] = igrf(time, latitude*180/pi, longitude*180/pi, altitude, 'geodetic');
    
    nedvec = [Bx_North,By_East,Bz_Down]; %nT, nT, nT
    ecivec = ned2eci(nedvec,latitude*180/pi,longitude*180/pi,dt);

    Bxi(stp,1) = ecivec(1); %nT
    Byi(stp,1) = ecivec(2); %nT
    Bzi(stp,1) = ecivec(3); %nT
    B_ECI_mag(stp,1) = sqrt(ecivec(1)^2 + ecivec(2)^2 + ecivec(3)^2);
    timey_wimey(stp,1) = dt;
end
%% specify static parameters
moment = 0.2;  % A m^2
sample_rate = 1;   % sec (1 Hz)

%% define mass properties
mass=10;     % mass in kg
height=0.34;   % height in m  (x)
width=0.2;   % width in m (y)
depth=0.1; % depth in m (z)
%
% rectangular prism spacecraft model
Jx=mass/12*(width^2+depth^2);
Jy=mass/12*(height^2+depth^2);
Jz=mass/12*(height^2+width^2);
%
% principal axis inertia tensor
J=[Jx,0,0;0,Jy,0;0,0,Jz];
%
% calculate coefficients
c1=(Jy-Jz)/Jx; d1=1/Jx;
c2=(Jz-Jx)/Jy; d2=1/Jy;
c3=(Jx-Jy)/Jz; d3=1/Jz;

%% controller
if cmode==1   % B-dot bang-bang control law
  % initialize parameters for calculation of B-dot
  % this section coded by Glenn Lightsey on February 11, 2022
  tlast=0; Bxlast=0; Bylast=0; Bzlast=0; mxlast=0; mylast=0; mzlast=0;
  %
  for i = 2:N
  %
    ti = tspan(i-1);
    hi = h(i-1);
    yi = Y(:,i-1);
  %
  % control algorithm goes here.
  % only calculate one control
  % distinguish between control sample steps and simulation steps
  %
    w1=yi(1); w2=yi(2); w3=yi(3);
    w=[w1;w2;w3];
    q1=yi(4); q2=yi(5); q3=yi(6); q4=yi(7);
    xi=[q4,-q3,q2;q3,q4,-q1;-q2,q1,q4;-q1,-q2,-q3];
    qd=0.5*xi*w;
  %
    C11=1-2*(q2^2+q3^2);
    C21=2*(q2*q1-q3*q4);
    C31=2*(q3*q1+q2*q4);
    C12=2*(q1*q2+q3*q4);
    C22=1-2*(q3^2+q1^2);
    C32=2*(q2*q3-q1*q4);
    C13=2*(q3*q1-q2*q4);
    C23=2*(q3*q2+q1*q4);
    C33=1-2*(q1^2+q2^2);
  %
    Bxeci1=Bxi(i-1);
    Byeci1=Byi(i-1);
    Bzeci1=Bzi(i-1);
    Bx_body=C11*Bxeci1+C12.*Byeci1+C13*Bzeci1;
    By_body=C21*Bxeci1+C22.*Byeci1+C23*Bzeci1;
    Bz_body=C31*Bxeci1+C32.*Byeci1+C33*Bzeci1;
  %
    Bvec_body=[Bx_body(1);By_body(1);Bz_body(1)]*1e-9; % units are in T
  %
    if (ti/sample_rate-floor(ti/sample_rate))<0.001, % test to see if this is a control update (1 msec test threshold)
      if (ti~=tlast),    % execute if this is not the first sample
        % update control
        % calculate new control
        Bdotx=(Bx_body(1)-Bxlast)/(ti-tlast); % units are in nanoT/s
        Bdoty=(By_body(1)-Bylast)/(ti-tlast);
        Bdotz=(Bz_body(1)-Bzlast)/(ti-tlast);
        mvec_body=-moment*[sign(Bdotx);sign(Bdoty);sign(Bdotz)]; % bang-bang control
        torque_body=cross(mvec_body,Bvec_body);
        mxlast=mvec_body(1);
        mylast=mvec_body(2);
        mzlast=mvec_body(3);
        % update BXxlast, Bylast, Bzlast, tlast for future reference
	    Bxlast=Bx_body(1);
	    Bylast=By_body(1);
	    Bzlast=Bz_body(1);
	    tlast=ti;
      else   % this is the first sample
        mvec_body=[0;0;0];     % zero control
	    torque_body=cross(mvec_body,Bvec_body);
        Bxlast=Bx_body(1);
        Bylast=By_body(1);
	    Bzlast=Bz_body(1);
	    tlast=ti;
      end
    else  % this is not a control sample
    % simulate with previous control
    % but torque is different, so need to calculate
      mvec_body=[mxlast;mylast;mzlast];
	  torque_body=cross(mvec_body,Bvec_body);
    end
    % pass torques to odefun
    % using global variables now, pass as function parameters possibly eventually
    L(1)=torque_body(1)*d1;
    L(2)=torque_body(2)*d2;
    L(3)=torque_body(3)*d3;
    LHIST(:,i)=L(1:3);   % store control torque in global history variable for plotting
    MHIST(:,i)=mvec_body(1:3); % store magnetic dipole in global variable for plotting
  %
  % integrator function calls happen below here
  % assumes control torque is the same, even though bfield changes at future times
  % if sample time is fast compared to change in bfield, should be okay.
  % to improve performance, calculate torque at each intermediate step
  % this is where you need interpolation step.
  %
    F(:,1) = feval(odefun,ti,yi,varargin{:});
    F(:,2) = feval(odefun,ti+0.5*hi,yi+0.5*hi*F(:,1),varargin{:});
    F(:,3) = feval(odefun,ti+0.5*hi,yi+0.5*hi*F(:,2),varargin{:});  
    F(:,4) = feval(odefun,tspan(i),yi+hi*F(:,3),varargin{:});
    Y(:,i) = yi + (hi/6)*(F(:,1) + 2*F(:,2) + 2*F(:,3) + F(:,4));
  end
end
%% this section coded by Erin McNeil on September 28, 2024
if cmode==2 % w cross b control
  % initialize parameters for calculation of B-dot
  % this section coded by Erin McNeil on September 28, 2024
  tlast=0; Bxlast=0; Bylast=0; Bzlast=0; mxlast=0; mylast=0; mzlast=0;
  %
  for i = 2:N
  %
    ti = tspan(i-1);
    hi = h(i-1);
    yi = Y(:,i-1);
  %
  % control algorithm goes here.
  % only calculate one control
  % distinguish between control sample steps and simulation steps
  %
    w1=yi(1); w2=yi(2); w3=yi(3);
    w=[w1;w2;w3];
    q1=yi(4); q2=yi(5); q3=yi(6); q4=yi(7);
    xi=[q4,-q3,q2;q3,q4,-q1;-q2,q1,q4;-q1,-q2,-q3];
    qd=0.5*xi*w;
  %
    C11=1-2*(q2^2+q3^2);
    C21=2*(q2*q1-q3*q4);
    C31=2*(q3*q1+q2*q4);
    C12=2*(q1*q2+q3*q4);
    C22=1-2*(q3^2+q1^2);
    C32=2*(q2*q3-q1*q4);
    C13=2*(q3*q1-q2*q4);
    C23=2*(q3*q2+q1*q4);
    C33=1-2*(q1^2+q2^2);
  %
    Bxeci1=Bxi(i-1);
    Byeci1=Byi(i-1);
    Bzeci1=Bzi(i-1);
    Bx_body=C11*Bxeci1+C12.*Byeci1+C13*Bzeci1;
    By_body=C21*Bxeci1+C22.*Byeci1+C23*Bzeci1;
    Bz_body=C31*Bxeci1+C32.*Byeci1+C33*Bzeci1;
  %
    Bvec_body=[Bx_body(1);By_body(1);Bz_body(1)]*1e-9; % units are in T
    
  %
    if (ti/sample_rate-floor(ti/sample_rate))<0.001 % test to see if this is a control update (1 msec test threshold)
      if (ti~=tlast)    % execute if this is not the first sample
        % update control
        % calculate new control
        Bdotx=(Bx_body(1)-Bxlast)/(ti-tlast); % units are in nanoT/s
        Bdoty=(By_body(1)-Bylast)/(ti-tlast);
        Bdotz=(Bz_body(1)-Bzlast)/(ti-tlast);

        B_body_norm = norm(Bvec_body);
        bx = Bx_body(1)/B_body_norm;
        by = By_body(1)/B_body_norm;
        bz = Bz_body(1)/B_body_norm;
        b = [bx;by;bz];
        k = 1;
        mvec_body = (k/B_body_norm)*cross(w,b);
        
        if mvec_body(1) > 0.2
            mvec_body(1) = 0.2;
        elseif mvec_body(1) < -0.2
            mvec_body(1) = -0.2;
        end
        
        if mvec_body(2) > 0.2
            mvec_body(2) = 0.2;
        elseif mvec_body(2) < -0.2
            mvec_body(2) = -0.2;
        end
        
        if mvec_body(3) > 0.2
            mvec_body(3) = 0.2;
        elseif mvec_body(3) < -0.2
            mvec_body(3) = -0.2;
        end
        
        torque_body=cross(mvec_body,Bvec_body);
        mxlast=mvec_body(1);
        mylast=mvec_body(2);
        mzlast=mvec_body(3);
        % update BXxlast, Bylast, Bzlast, tlast for future reference
	    Bxlast=Bx_body(1);
	    Bylast=By_body(1);
	    Bzlast=Bz_body(1);
	    tlast=ti;
      else   % this is the first sample
        mvec_body=[0;0;0];     % zero control
	    torque_body=cross(mvec_body,Bvec_body);
        Bxlast=Bx_body(1);
        Bylast=By_body(1);
	    Bzlast=Bz_body(1);
	    tlast=ti;
      end
    else  % this is not a control sample
    % simulate with previous control
    % but torque is different, so need to calculate
      mvec_body=[mxlast;mylast;mzlast];
	  torque_body=cross(mvec_body,Bvec_body);
    end
    % pass torques to odefun
    % using global variables now, pass as function parameters possibly eventually
    L(1)=torque_body(1)*d1;
    L(2)=torque_body(2)*d2;
    L(3)=torque_body(3)*d3;
    LHIST(:,i)=L(1:3);   % store control torque in global history variable for plotting
    MHIST(:,i)=mvec_body(1:3); % store magnetic dipole in global variable for plotting
  %
  % integrator function calls happen below here
  % assumes control torque is the same, even though bfield changes at future times
  % if sample time is fast compared to change in bfield, should be okay.
  % to improve performance, calculate torque at each intermediate step
  % this is where you need interpolation step.
  %
    F(:,1) = feval(odefun,ti,yi,varargin{:});
    F(:,2) = feval(odefun,ti+0.5*hi,yi+0.5*hi*F(:,1),varargin{:});
    F(:,3) = feval(odefun,ti+0.5*hi,yi+0.5*hi*F(:,2),varargin{:});  
    F(:,4) = feval(odefun,tspan(i),yi+hi*F(:,3),varargin{:});
    Y(:,i) = yi + (hi/6)*(F(:,1) + 2*F(:,2) + 2*F(:,3) + F(:,4));
  end
end
Y = Y.';
